Introduce a 'K' type param to typed TimerScheduler

dwijnand opened the pull request.

dwijnand committed:
No need to cast ActorContext, just pass the log

dwijnand committed:
Introduce a 'K' type param to typed TimerScheduler

dwijnand committed:
Fix typed testkit test compilation

dwijnand committed:
Fix typed testkit tests test compilation

hepin1989 commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/TimerSchedulerImpl.scala:
```diff
@@ -18,55 +18,52 @@ import scala.concurrent.duration.FiniteDuration
-  final case class InfluenceReceiveTimeoutTimerMsg(key: Any, generation: Int, owner: AnyRef) extends TimerMsg
-  final case class NotInfluenceReceiveTimeoutTimerMsg(key: Any, generation: Int, owner: AnyRef)
-    extends TimerMsg with NotInfluenceReceiveTimeout
+  final case class Timer[K, T](key: K, msg: T, repeat: Boolean, generation: Int, task: Cancellable)
```
`Key`

dwijnand committed:
Allow nesting timers in TimerSpec

dwijnand committed:
Introduce TimerSchedulerImpl count to fix discarding timers

dwijnand committed:
Fix persistence typed test compilation

dwijnand committed:
Go back to TimerMsg using TimerSchedulerImpl as a reference

Lose isSame, again, but it probably wasn't ever being triggered.

dwijnand committed:
Restore TimerMsg#generation too, for multiple timer sets

dwijnand committed:
Introduce Behaviours.withTypedTimers to avoid breaking easy type inferrence

dwijnand committed:
Finish fixing akka-persistence-typed test compilation

dwijnand committed:
No longer need ActorContext#timerCount

dwijnand committed:
Restore stack-saving nesting, for untyped timer keys

dwijnand committed:
Dedup in TimerSchedulerImpl

dwijnand committed:
Rename to key-typed timers

dwijnand committed:
Generalise TimerSchedulerImpl.wrapWithTimers

dwijnand committed:
Restore TimerScheduler's original type constructor

... by introducing KeyTypedTimerScheduler & AnyKeyTimerSchedulerImpl

dwijnand committed:
Fix TimerSpec nesting test

dwijnand committed:
Remove a no-longer-required change in the aggregate diff

dwijnand committed:
Bring back TimerInterceptor#isSame

dwijnand committed:
Inline TimerSchedulerImpl#isSame

Don'\''t specialise for AnyKeyTimerSchedulerImpl, checking for eq is
enough _and_ is safer.

dwijnand committed:
Rename to SharedSchedulerImpl

dwijnand committed:
Fix TimerInterceptor#isSame, duuurrr

dwijnand committed:
Oh wait, just have one TimerSchedulerImpl, with all keys

dwijnand committed:
Recover the TimerSchedulerImpl name

dwijnand committed:
Correct comment

dwijnand committed:
Fix log used in interceptTimerMsg

dwijnand committed:
Replace withKeyTypedTimers with withKeyType

dwijnand committed:
Drop the key type param on inner types

Turns out they're not used

dwijnand committed:
Make TimerInterceptor take just a TimerSchedulerImpl

dwijnand committed:
Add withKeyType to javadsl.TimerScheduler too

dwijnand committed:
Move intercept to TimerSchedulerImpl

dwijnand committed:
Make the key type param contravariant, eliminate unnecessary casts

dwijnand committed:
Replace TimerMsg hierarchy with a simple mix-in

dwijnand commented:
OK, I'm fairly sure I'm done.

The last thing I would want to do is make `interceptTimerMsg` return an `OptionVal` rather than `null`, but I need to make `OptionVal` globally generic, which is in https://github.com/akka/akka/pull/25863/commits/b7811702f35b71ece08f44acbbd02a3c6a560a4f.  I can do it in a follow-up PR.

What do you think?

dwijnand commented:
Oh and I'm good with git's, so I'm happy to shuffle changes, split PRs, or whatever.  Just let me know what you need.

dwijnand committed:
Re-implement using a type member instead

This commit is more experimental..

dwijnand committed:
Add a type-checking test for withKeyType

dwijnand committed:
Revert "Re-implement using a type member instead"

This reverts commit c09016b56f2ddb1e4f13d803bb1ed5fdab611d99.

Interesting idea, but there are no type members in Java, so this
wouldn't have feature parity.

chbatey commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala:
```diff
@@ -308,8 +308,16 @@ class TimerSpec extends ScalaTestWithActorTestKit(
+
+    "typecheck when using withKeyType" in {
+      Behaviors.withTimers[String] { timers0 ⇒
+        val timers = timers0.withKeyType[String]
```
was there a mistake you made with the old API that motivated this change? I'm not sure I see the value but if others do then happy to merge

dwijnand replied:
Personally, I haven't made this mistake yet.  But there's an easy risk of using the wrong key type, breaking your program.  For instance, if you have a reasonably complex actor, where you're passing the timers instance around various behaviour-defining methods, like one of the ones in the examples of timers.

patriknw replied:
I'm not sure it's worth the additional API complexity. If we think it's important we should enforce specifying the key type in `withTimers` rather than having a default Any and having to opt in to something safer (probably only found after doing the mistake)

dwijnand replied:
The downside of that is you lose some nice type inference when wrapping, because it needs you to define the type of the key, and therefore you also have to define the type of message... 😕

So you go from

```scala
Behaviors.withTimers { timers =>
  Behaviors.same[Message]
}
```

to

```scala
Behaviors.withTimers[String, Message] { timers =>
  Behaviors.same[Message]
}
```

Or we could do some of those shenanigans that allow for "currying type parameters"...

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/TimerScheduler.scala:
```diff
@@ -15,7 +15,7 @@ import java.time.Duration
  * the actor that owns it.
  */
-trait TimerScheduler[T] {
+trait KeyTypedTimerScheduler[-K, T] {
```
Not sure the rename is worth it being a breaking change for a public type even if it is apimaychange

dwijnand replied:
The class `TimerScheduler` is preserved, as a subclass of this one.  So no binary breakage, verified for Scala 2.12 with MiMa: https://github.com/lightbend/migration-manager/pull/262.

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/TimerScheduler.scala:
```diff
@@ -53,11 +53,17 @@ trait TimerScheduler[T] {
+
+trait TimerScheduler[T] extends KeyTypedTimerScheduler[Any, T] {
+
+  def withKeyType[K]: KeyTypedTimerScheduler[K, T] = this
```
scaladoc

dwijnand commented:
So I'm going to resubmit this change, cleaned up and including the missing scaladoc.

I've also got a version of this I've been trying, using a type member rather than a type parameter but it still requires introducing 2 `asInstanceOf`s, which is disappointing.

dwijnand commented:
Cleaned up, rebased, updated and resubmitted as two alternatives:

* #25878 Introduce KeyTypedTimerScheduler to type timers keys, and
* #25880 Introduce type member K to TimerScheduler to type timer keys
